pallets/settlement/src/lib.rs

```diff
@@ -1541,4 +1544,25 @@ impl<T: Trait> Module<T> {
         }
         Ok(())
     }
+
+    /// Returns total number of legs of an `instruction_id` and vector of legs where sender is in the `portfolios` set.
+    /// Also, ensures that the number of filtered legs is under the limit.
+    fn filtered_legs(
+        instruction_id: u64,
+        portfolios: &BTreeSet<PortfolioId>,
+        max_filtered_legs: u32,
+    ) -> Result<(u32, Vec<(u64, Leg<T::Balance>)>), DispatchError> {
+        let legs = <InstructionLegs<T>>::iter_prefix(instruction_id).collect::<Vec<_>>();
+        let legs_count = legs.len() as u32;
+        let filtered_legs = legs
+            .into_iter()
+            .filter(|(_, leg_details)| portfolios.contains(&leg_details.from))
+            .collect::<Vec<_>>();
```

Comment:
```suggestion
        let mut legs_count = 0;
        let filtered_legs = <InstructionLegs<T>>::iter_prefix(instruction_id)
            .into_iter()
            .inspect(|_| legs_count += 1)
            .filter(|(_, leg_details)| portfolios.contains(&leg_details.from))
            .collect::<Vec<_>>();
```

Reply:
Honestly, I don't think this is an upgrade. Sure, it reduces the memory footprint by a bit but it makes the iteration a tad bit more complex and requires mutable variables.